### DISCUSSION TITLE
I can't find a way to apply a plugin only to specific store.

### DISCUSSION THREAD

**Discussion by fermaud** [vuejs/pinia#2027] [Help and Questions] [answered]
In the doc, it says `Apply only to specific stores`
Is there any option to do that ? I could not find a way to do it. 
https://pinia.vuejs.org/core-concepts/plugins.html#plugins

**Comment by ilyarefraction** [comment, marked as answer]
Hm, if I get you clear you just need to compare `store.$id` with id of the store you need to patch.
Like this:
```js
const id = 'yourId'
pinia.use(({ store }) => {
  if(store.$id === id) {
     ....
})
```

**Reply by fermaud** [reply]
Thank your for the answer @maIIady , the problem with this solution is that you will still have typescript autocompletion of the plugin methods even for stores where you don't need the plugin.

**Reply by posva** [reply]
Unfortunately, TS cannot infer the type of the store here, you should be able to get the type of the store with

```ts
const storeTyped = store as ReturnType<typeof useYourIdStore>
```

**Reply by Shiva127** [reply]
In this case you can write:
```ts
import { useStore } from '@/stores/store

pinia.use(({ store }) => {
  if(store.$id === useStore.$id) {
    const storeTyped = store as ReturnType<typeof useStore>
    ...
  }
})
```

**Comment by isimmons** [comment]
This method would not work for me at all. to load an imported plugin. But I used the same conditional at the top of the actual plugin for an early return and it works.

As for Typescript autocompletion I have the opposite problem. Even before restricting the plugin to the one store, the plugin methods were missing from the stores type. I had to create a custom type in the template where I want to use the plugin methods.

```js
import { useProductStore, useCartStore } from '~/stores';

const productStore = useProductStore();
const cartStore = useCartStore();

export type CartStoreWithPlugin = typeof cartStore & {
  undoHistory: () => void;
  redoHistory: () => void;
};

const store = cartStore as CartStoreWithPlugin;
```

**Comment by alwin** [comment]
Found this page via search on the web so leaving my solution here for others.

To use a specific store inside the pinia plugin, you can create a type guard, to avoid any `as` cast inside the plugin body. 

```ts
import type { PiniaPlugin } from "pinia";
import type { useMyStore } from "./myStore";

type MyStore = ReturnType<typeof useMyStore>;

// TS type guard
function isMyStore(store: { $id: string }): store is MyStore {
  return store.$id === "my-store";
}

const myPlugin: PiniaPlugin = ({ store }) => {
  if (!isMyStore(store)) return;

  // store is fully typed as MyStore, no cast needed
  store.myAction();
};
```

Thanks for the earlier answers! They helped so I could make it work.
